prov/sm2/src/sm2_coordination.c

```diff
@@ -0,0 +1,574 @@
+	struct sm2_ep_allocation_entry *entries;
+	int fd, common_fd, err, tries, item;
+	bool have_file_lock = false;
+
+	sprintf(template, "%s/fi_sm2_pid%d_XXXXXX", SM2_COORDINATION_DIR,
+		getpid());
+
+	/* Assume we are the first process here.
+	 * Try to open a tmpfile file in the shm directory
+	 */
+	fd = mkostemp(template, O_RDWR);
+	assert(fd > 0);
+	err = ftruncate(fd, sizeof(*header));
+	if (OFI_UNLIKELY(err))
+		return -FI_ENOMEM;
+
+	header = sm2_mmap_map(fd, &map_ours);
+
+	pthread_mutexattr_init(&att);
+	pthread_mutexattr_setpshared(&att, PTHREAD_PROCESS_SHARED);
+	pthread_mutex_init(&header->write_lock, &att);
+
+	header->file_version = SM2_VERSION;
+	header->ep_region_size = sm2_calculate_size_offsets(NULL, NULL);
+	header->ep_allocation_offset = sizeof(*header);
+	header->ep_allocation_offset = NEXT_MULTIPLE_OF(
+		header->ep_allocation_offset, SM2_MAX_UNIVERSE_SIZE);
+	header->ep_regions_offset = header->ep_allocation_offset +
+				    (SM2_MAX_UNIVERSE_SIZE * sizeof(*entries));
+	header->ep_regions_offset = NEXT_MULTIPLE_OF(header->ep_regions_offset,
+						     SM2_MAX_UNIVERSE_SIZE);
+
+	/* allocate enough space in the file for all our allocations, but no
+	 * data exchange regions yet.
+	 */
+	sm2_mmap_remap(&map_ours, header->ep_regions_offset);
+	header = (struct sm2_coord_file_header *) map_ours.base;
+	entries = sm2_mmap_entries(&map_ours);
+	for (item = 0; item < SM2_MAX_UNIVERSE_SIZE; item++)
+		entries[item].pid = 0;
+
+	pthread_mutex_lock(&header->write_lock);
+
+	/* Make sure the header is written before we link the file,
+	 * flush file
+	 */
+	atomic_mb();
+
+	tries = SM2_STARTUP_MAX_TRIES;
+	do {
+		/* create a hardlink to our file with the common name.
+		 *  - on success: we hold the lock to a newly
+		 *		  created coordination file.
+		 *  - on failure: the file already exists, we should try opening
+		 *  		  it.
+		 */
+		err = link(template, SM2_COORDINATION_FILE);
+		if (0 == err) {
+			/* We are using the memory we initialized.  Duplicate it
+			 * and leave the map open. */
+			memcpy(map_shared, &map_ours, sizeof(map_ours));
+			have_file_lock = true;
+			break;
+		}
+
+		common_fd = open(SM2_COORDINATION_FILE, O_RDWR);
+		if (common_fd > 0) {
+			tmp_header = sm2_mmap_map(common_fd, map_shared);
+			assert(map_shared->size >= sizeof(*tmp_header));
+			err = pthread_mutex_trylock(&tmp_header->write_lock);
+			if (err == 0) {
+				sm2_file_attempt_shrink(map_shared);
+				sm2_mmap_cleanup(&map_ours);
+				have_file_lock = true;
+				break;
+			}
+
+			/* TODO Handle PID that dies while holding file lock */
+
+			sm2_mmap_cleanup(map_shared);
+		}
+		/* we could not acquire the lock, sleep and try again. */
+		/* TODO Consider reducing the amount of time we sleep here */
+		usleep(10000);
+	} while (tries-- > 0);
+
+	unlink(template);
+
+	if (!have_file_lock) {
+		FI_WARN(&sm2_prov, FI_LOG_AV,
+			"Failed to acquire the lock to the coordination file. "
+			"Aborting\n");
+		return -FI_EAVAIL;
```

Comment:
We should be holding a lock here, which needs to be released before returning.

Reply:
No.  The way open_and_create() works is it assumes that every joining peer will be the first to create the global SHM file, and they name that file with their PID in the name.  The user than initializes the file, and grabs the lock on the new file they created (that only they are using).  If we can successfully link our file to the "global named file", that means we were the first to get there, else our file will get unlinked (removed)... and that lock that we initialized in it doesn't matter b/c nobody will ever use that file again.